File: C++/jwcfg/JWParser.cpp
#include "JWParser.h"

void AddToStringList(const std::string &str, std::list<std::string> &strList)
{
	int equalCount = CharaterCount(str.c_str(), str.length(), '=');

	if (equalCount == 1)
	{
		strList.push_back(str);
	}
	else if (equalCount > 1) // Only possible if a string...
	{
		int quoteCount = CharaterCount(str.c_str(), str.length(), '"');

		if (quoteCount >= 2)
			strList.push_back(str); // Assume valid string for now
	}
	else
	{
		printf("Skipping this malformed JW key-value pair: %s\n", str.c_str());
	}
}

JWParser::EReturnCode JWParser::Parse(const Buffer &buffer, Map &jwMap)
{
	std::list<std::string> strPairs;
	std::string str;

	const u8 *buf = buffer.get();

	for (u32 i = 0; i < buffer.getSize(); ++i)
	{
		c8 ch = (c8)buf[i];

		if (IsNewLine(ch) || IsEOF(ch))
		{
			if (ch == '\r') // Skip past \r,
				continue;

			if (str.length() > 0)
			{
				AddToStringList(str, strPairs);
				str.clear(); // No matter what, clear for next go around
			} 
		}
		else
		{
			c8 s[2] = { ch, 0 };
			str.append(s);
		}
	}

	// Check to see if final string needs to be added!
	if (str.length() > 0)
	{
		AddToStringList(str, strPairs);
		str.clear(); // No matter what, clear for next go around
	} 

	if (strPairs.size() == 0)
		return EReturnCode::eNoValidData;

	jwMap.clear(); // Clear out anything in map, ready to fill with new values

	do
	{
		const std::string &s = strPairs.back();
		
		if (ParseKeyValue(s, jwMap) != EReturnCode::eSuccess)
		{
			printf("Unable to parse following key-value pair: %s\n", s.c_str());			
		}

		strPairs.pop_back();
	}
	while(strPairs.size() > 0);

	return EReturnCode::eSuccess;
}

JWParser::EReturnCode JWParser::ParseKeyValue(const std::string &str, Map &jwMap)
{
	const c8 *s = str.c_str();
	int len = (int)str.length();

	const int equalIdx = FindFirstOf(s, (u32)len, '=');

	if (equalIdx == -1)
		return EReturnCode::eInvalidKey;
	
	// Remove leading whitespace
	int sIdx = 0;

	while((sIdx < len) && (IsWhiteSpace(s[sIdx])))
	{
		++sIdx;
	}

	if (sIdx >= equalIdx)
		return EReturnCode::eInvalidKey;

	// Remove trailing whitespace
	int eIdx = equalIdx;

	while((eIdx > 1) && (IsWhiteSpace(s[eIdx - 1])))
	{
		--eIdx;
	}

	if (eIdx < sIdx)
		return EReturnCode::eInvalidKey;

	std::string key(s + sIdx, eIdx - sIdx);

	// Remove leading whitespace
	sIdx = equalIdx + 1;
	
	while((sIdx < len) && (IsWhiteSpace(s[sIdx])))
		++sIdx;

	if (sIdx >= len)
		return EReturnCode::eUnexpectedValue;

	// Remove trailing whitespace
	eIdx = len;

	while((eIdx > equalIdx) && (IsWhiteSpace(s[eIdx - 1])))
	{
		--eIdx;
	}

	if (eIdx < sIdx)
		return EReturnCode::eInvalidKey;

	std::string value(s + sIdx, eIdx - sIdx);

	Value jwVal;
	
	JWParser::EReturnCode rc = ParseValue(value, jwVal);

	if (rc != EReturnCode::eSuccess)
		return rc;
	
	jwMap.insert(Pair(key, jwVal));
		return EReturnCode::eSuccess;
}

JWParser::EReturnCode JWParser::ParseValue(const std::string &str, Value &jwValue)
{
	const c8 *s = str.c_str();
	u32 len = str.length();
	c8 ch = s[0];

	// Possibily an integer or float
	if (IsDigit(ch) || ch == '-' || ch == '.')
	{
		bool foundPeriod = (ch == '.');

		if (!foundPeriod)
		{
			int idx = FindFirstOf(s, len, '.');

			if (idx != -1)
				foundPeriod = true;
		}

		if (foundPeriod) // It's a float
		{
			float f = (float)atof(s);
			jwValue.set(f);
		}
		else // It's a int
		{
			int i = atoi(s);
			jwValue.set(i);
		}

		return EReturnCode::eSuccess;
	}
	// Possibly a bool
	else if (ch == 't' || ch == 'T' || ch == 'f' || ch == 'F')
	{
		if (len == 4) // Check for truth
		{
			bool R = (s[1] == 'r' || s[1] == 'R');
			bool U = (s[2] == 'u' || s[2] == 'U');
			bool E = (s[3] == 'e' || s[3] == 'E');

			if (R && U && E)
			{
				jwValue.set(true);
					return EReturnCode::eSuccess;
			}
			else
				return EReturnCode::eMalformedData;
		}
		else if (len == 5) // Check for falsehood
		{
			bool A = (s[1] == 'a' || s[1] == 'A');
			bool L = (s[2] == 'l' || s[2] == 'L');
			bool S = (s[3] == 's' || s[3] == 'S');
			bool E = (s[4] == 'e' || s[4] == 'E');

			if (A && L && S && E)
			{
				jwValue.set(false);
					return EReturnCode::eSuccess;
			}
			else
				return EReturnCode::eMalformedData;
		}
	}
	// Possibly a string
	else if (ch == '"')
	{
		if (str[len - 1] == '"')
		{
			std::string subStr(str, 1, len - 2); 
			jwValue.set(subStr);
				return EReturnCode::eSuccess;
		}
	}

	return EReturnCode::eMalformedData;
}

File: C#/jwcfg/JWParser.cs
using System.Collections.Generic;
using System.IO;
using System.Text;

static class JWParser
{
	public sealed class Value
	{	
		public enum EType
		{
			eEmpty = 0,
			eInt,
			eFloat,
			eBool,
			eString
		};

		public Value()
		{
			mType = EType.eEmpty;
			mInt = 0;
			mFloat = 0.0f;
			mBool = false;
		}

		public void set(int i) { mType = EType.eInt; mInt = i; }
		public void set(float f) { mType = EType.eFloat; mFloat = f; }
		public void set(bool b) { mType = EType.eBool; mBool = b; }
		public void set(in string s) { mType = EType.eString; mString = s; }

		public EType getType() { return mType; }
		public int asInt() { return mInt; }
		public float asFloat() { return mFloat; }
		public bool asBool() { return mBool; }
		public ref string asString() { return ref mString; }

		EType mType;
		int mInt;
		float mFloat;
		bool mBool;
		string mString;
	}

     public enum EReturnCode
     {
          eSuccess = 0,
		eInvalidKey,
		eMalformedData,
		eNoValidData,
		eUnexpectedValue
     };

     public static EReturnCode Parse(in string jwBuffer, ref Dictionary<string, Value> jwMap)
	{
		MemoryStream memStream = new MemoryStream(Encoding.UTF8.GetBytes(jwBuffer));
		List<string> strLines = new List<string>();

		using (StreamReader reader = new StreamReader(memStream))
		{
			while (!reader.EndOfStream)
			{
				var s = reader.ReadLine();

				if (IsValid(s))
					strLines.Add(s);
			}
		}

		if (strLines.Count == 0)
			return EReturnCode.eNoValidData;

		jwMap.Clear();

		for (int i = 0; i < strLines.Count; ++i)
		{
			var s = strLines[i];
			int equalIdx = s.IndexOf('=');
			
			if (equalIdx == -1)
				continue;
			
			string keyString = s.Substring(0, equalIdx).Trim();
			string valueString = s.Substring(equalIdx + 1).Trim();

			var v = new Value();

			JWParser.EReturnCode rc = ParseValue(valueString, ref v);

			if (rc != EReturnCode.eSuccess)
				return rc;

			jwMap.Add(keyString, v);
		}

		return EReturnCode.eSuccess;
	}

	public static EReturnCode Parse(string[] linesOfText, ref Dictionary<string, Value> jwMap)
	{
		if (linesOfText == null)
			return EReturnCode.eNoValidData;

		if (linesOfText.Length == 0)
			return EReturnCode.eNoValidData;

		jwMap.Clear();

		for (int i = 0; i < linesOfText.Length; ++i)
		{
			var s = linesOfText[i];
			int equalIdx = s.IndexOf('=');
			
			if (equalIdx == -1)
				continue;
			
			string keyString = s.Substring(0, equalIdx).Trim();
			string valueString = s.Substring(equalIdx + 1).Trim();

			var v = new Value();

			JWParser.EReturnCode rc = ParseValue(valueString, ref v);

			if (rc != EReturnCode.eSuccess)
				return rc;

			jwMap.Add(keyString, v);
		}

		return EReturnCode.eSuccess;
	}

	public static string ToString(Dictionary<string, Value> jwMap)
	{
		if (jwMap.Count == 0)
			return string.Empty;

		var strBuilder = new StringBuilder();

		foreach (var entry in jwMap)
		{
			string line = string.Empty;

			switch(entry.Value.getType())
			{
				case Value.EType.eBool: line = string.Format("{0}={1}", entry.Key, entry.Value.asBool()); break;
				case Value.EType.eInt: line = string.Format("{0}={1}", entry.Key, entry.Value.asInt()); break;
				case Value.EType.eString: line = string.Format("{0}=\"{1}\"", entry.Key, entry.Value.asString()); break;
				case Value.EType.eFloat: line = string.Format("{0}={1}", entry.Key, entry.Value.asFloat()); break;
				default: break;
			}

               strBuilder.AppendLine(line);
          }

		return strBuilder.ToString();
	}

	private static bool IsValid(in string line)
	{
		int eq1 = line.IndexOf('=');
		int eq2 = line.LastIndexOf('=');

		// No equal signs, not valid
		if (eq1 == -1)
			return false;

		// Exactly 1 equal sign, valid
		if (eq1 == eq2)
			return true; 

		// Only way to have 2 or more equal signs is if it is a string.... Check for string formating
		int qIdx = line.Length - 1;

		// First check end of string
		if (line[qIdx] != '"')
			return false;
		else // Need to see if first non-white char is a '"' before next '='
		{
			qIdx = eq1 + 1;
			char ch = ' ';

			while (qIdx < eq2)
			{
				ch = line[qIdx];

				if (char.IsWhiteSpace(ch))
					qIdx++;
				else
					break;
			}

			return ch.Equals('"');
		}		
	}

	static EReturnCode ParseValue(in string str, ref Value jwValue)
	{
		char ch = str[0];

		// Possibily an integer or float
		if (char.IsDigit(ch) || ch == '-' || ch == '.')
		{
			bool foundPeriod = (ch == '.');

			// Look for a period
			if (!foundPeriod)
			{
				int idx = str.IndexOf('.');
				
				if (idx != -1)
					foundPeriod = true;
			}
			
			if (foundPeriod) // It's a float
			{
				float f = 0.0f;
				
				if (float.TryParse(str, out f) == false)
					return EReturnCode.eMalformedData;

				jwValue.set(f);
			}
			else // It's a int
			{
				int i = 0;

				if (int.TryParse(str, out i) == false)
					return EReturnCode.eMalformedData;

				jwValue.set(i);
			}

			return EReturnCode.eSuccess;
		}
		// Possibly a bool
		else if (ch == 't' || ch == 'T' || ch == 'f' || ch == 'F')
		{
			if (str.Length == 4) // Check for truth
			{
				bool R = (str[1] == 'r' || str[1] == 'R');
				bool U = (str[2] == 'u' || str[2] == 'U');
				bool E = (str[3] == 'e' || str[3] == 'E');

				if (R && U && E)
				{
					jwValue.set(true);
						return EReturnCode.eSuccess;
				}
				else
					return EReturnCode.eMalformedData;
			}
			else if (str.Length == 5) // Check for falsehood
			{
				bool A = (str[1] == 'a' || str[1] == 'A');
				bool L = (str[2] == 'l' || str[2] == 'L');
				bool S = (str[3] == 's' || str[3] == 'S');
				bool E = (str[4] == 'e' || str[4] == 'E');

				if (A && L && S && E)
				{
					jwValue.set(false);
						return EReturnCode.eSuccess;
				}
				else
					return EReturnCode.eMalformedData;
			}
		}
		// Check for string
		else if (str[0] == '"' && str[str.Length - 1] == '"' && str.Length >= 3)
		{
			string s = str.Substring(1, str.Length - 2);
			jwValue.set(s);
				return EReturnCode.eSuccess;
		}

		return EReturnCode.eMalformedData;
	}
}